### PR TITLE
MODDCB-49

### PR DESCRIPTION
**Pull request by MagzhanArtykov**
Covers: [MODDCB-49](https://issues.folio.org/browse/MODDCB-49)

## Purpose
Main purpose of this PR is to implement support of DCB transaction cancellation

## Approach
If a user cancels DCB Page request in Lending library, then the status of DCB transaction should be updated to CANCELLED.
DCB propagates CANCELLED transaction status to Borrowing/Pickup libraries.

## Learning
Please refer to: [CIRCSTORE-463](https://issues.folio.org/browse/CIRCSTORE-463)

### CODE REVIEW

File: descriptors/ModuleDescriptor-template.json [hunk 1]
```diff
@@ -91,7 +91,9 @@
             "circulation-item-storage.items.item.post",
             "circulation-item-storage.items.item.get",
             "inventory-storage.material-types.collection.get",
-            "inventory-storage.loan-types.collection.get"
+            "inventory-storage.loan-types.collection.get",
```
Comment (Vignesh-kalyanasundaram) [inline descriptors/ModuleDescriptor-template.json:94]: I think we may need the below permissions as well
**circulation-storage.requests.item.get**
**circulation.requests.item.get**
Please check

File: src/main/java/org/folio/dcb/service/impl/CirculationRequestServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+@RequiredArgsConstructor
+public class CirculationRequestServiceImpl implements CirculationRequestService {
+
+  private final CirculationRequestClient circulationStorageClient;
+  private final FolioExecutionContext folioExecutionContext;
+
+  @Override
+  public CirculationRequest fetchRequestById(String requestId) {
+    log.info("fetchRequestById:: fetching request for id {} ", requestId);
+    return circulationStorageClient.fetchRequestById(requestId);
```
Comment (DmytroBykov1) [inline src/main/java/org/folio/dcb/service/impl/CirculationRequestServiceImpl.java:27]: as far as I see, the request storage returns 404 in the case of Item with a given ID not found. so somewhere should be implemented FeignException.NotFound processing

**Review by DmytroBykov1 — CHANGES_REQUESTED** [review]

File: src/main/java/org/folio/dcb/listener/kafka/CirculationEventListener.java [hunk 3]
```diff
@@ -82,4 +86,29 @@ public void handleCheckOutEvent(String data, MessageHeaders messageHeaders) {
...
+    String tenantId = getHeaderValue(messageHeaders, XOkapiHeaders.TENANT, null).get(0);
+    var eventData = parseEvent(data);
+    if (Objects.nonNull(eventData) && eventData.getType() == EventData.EventType.CANCEL) {
+      String itemID = eventData.getItemId();
+      if (Objects.nonNull(itemID)) {
+        log.info("updateTransactionStatus:: Received cancel event for itemId: {}", itemID);
+        systemUserScopedExecutionService.executeAsyncSystemUserScoped(tenantId, () ->
+          transactionRepository.findTransactionByItemIdAndStatusNotInClosed(UUID.fromString(itemID))
+            .ifPresent(transactionEntity -> {
+              switch (transactionEntity.getRole()) {
```
Comment (Vignesh-kalyanasundaram) [inline src/main/java/org/folio/dcb/listener/kafka/CirculationEventListener.java:104]: As the implementation is same for all the roles, I think we don't need to use case and default here. we may call like below.
```suggestion
                  systemUserScopedExecutionService.executeAsyncSystemUserScoped(tenantId, () ->
          transactionRepository.findTransactionByItemIdAndStatusNotInClosed(UUID.fromString(itemID))
            .ifPresent(baseLibraryService::cancelTransactionEntity)
        );
```

File: src/main/java/org/folio/dcb/service/CirculationRequestService.java [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package org.folio.dcb.service;
+
+import org.folio.dcb.domain.dto.CirculationRequest;
+
+public interface CirculationRequestService {
+  CirculationRequest fetchRequestById(String requestId);
```
Comment (Vignesh-kalyanasundaram) [inline src/main/java/org/folio/dcb/service/CirculationRequestService.java:6]: I see fetchRequestById is not being called anywhere outside the class. Do you think we need to have it here

File: src/main/java/org/folio/dcb/service/impl/CirculationRequestServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+@Service
+@Log4j2
+@RequiredArgsConstructor
+public class CirculationRequestServiceImpl implements CirculationRequestService {
+
+  private final CirculationRequestClient circulationStorageClient;
+  private final FolioExecutionContext folioExecutionContext;
+
+  @Override
+  public CirculationRequest fetchRequestById(String requestId) {
```
Comment (Vignesh-kalyanasundaram) [inline src/main/java/org/folio/dcb/service/impl/CirculationRequestServiceImpl.java:25]: If you agree with the comment provided on the circulationRequestService, we can make this method private

File: src/main/resources/db/changelog/changes/create-transaction-table.xml [hunk 6]
```diff
@@ -13,6 +12,14 @@
     </sql>
   </changeSet>
 
+  <!-- Create StatusEnum ENUM -->
```
Comment (Vignesh-kalyanasundaram) [inline src/main/resources/db/changelog/changes/create-transaction-table.xml:15]: As the code is already deployed in testing environment, Please add this in a new xml file and add it to changelog file.

File: src/main/resources/db/changelog/changes/create-transaction-table.xml [hunk 7]
```diff
@@ -85,6 +92,9 @@
       <column name="borrowing_library_code" type="text">
         <constraints nullable="false"/>
       </column>
+      <column name="request_id" type="UUID">
```
Comment (Vignesh-kalyanasundaram) [inline src/main/resources/db/changelog/changes/create-transaction-table.xml:95]: Please move this as well

**Review by Vignesh-kalyanasundaram — CHANGES_REQUESTED** [review]

File: src/main/java/org/folio/dcb/service/impl/CirculationRequestServiceImpl.java (hunk 2 above)
Comment (MagzhanArtykov) [reply src/main/java/org/folio/dcb/service/impl/CirculationRequestServiceImpl.java:27]: Accepted

File: src/main/java/org/folio/dcb/service/impl/CustomTenantService.java [hunk 8]
```diff
@@ -235,4 +240,17 @@ private void createHolding() {
...
+
+  private void createCancellationReason(){
+    try {
+      cancellationReasonClient.findCancellationReason(CANCELLATION_REASON_ID);
+    } catch (FeignException.NotFound ex) {
+      log.debug("createCancellationReason:: creating cancellation reason");
+      cancellationReasonClient.createCancellationReason(CancellationReasonClient.CancellationReason.builder()
+        .id(CANCELLATION_REASON_ID)
+        .description("DCB Cancelled")
+        .name("DCB Cancelled").build());
```
Comment (gurleenkaurbp) [inline src/main/java/org/folio/dcb/service/impl/CustomTenantService.java:252]: Prefer a string constant for "DCB Cancelled" value.

File: src/main/java/org/folio/dcb/service/impl/LendingLibraryServiceImpl.java [hunk 9]
```diff
@@ -65,6 +66,10 @@ public void updateTransactionStatus(TransactionEntity dcbTransaction, Transactio
       updateTransactionEntity(dcbTransaction, requestedStatus);
     } else if (ITEM_CHECKED_OUT == currentStatus && ITEM_CHECKED_IN == requestedStatus) {
       updateTransactionEntity(dcbTransaction, requestedStatus);
+    } else if(CANCELLED == requestedStatus) {
```
Comment (gurleenkaurbp) [inline src/main/java/org/folio/dcb/service/impl/LendingLibraryServiceImpl.java:69]: @MagzhanArtykov  - Where is the code or where will the check happen for the below mentioned lines ? 
The cancellation request can be applied to DCB transactions that are in the following statuses [CREATED, OPEN, AWAITING_PICKUP]. For the transactions that are in status [ITEM_CHECKED_OUT, ITEM_CHECKED_IN] FOLIO should reject such requests with a "400 Bad request" response.

Comment (gurleenkaurbp) [reply src/main/java/org/folio/dcb/service/impl/LendingLibraryServiceImpl.java:70]: The check will happen on Front-End. Therefore, resolving this conversation.

File: src/main/java/org/folio/dcb/service/impl/BaseLibraryService.java [hunk 10]
```diff
@@ -102,13 +106,31 @@ public void updateTransactionStatus(TransactionEntity dcbTransaction, Transactio
...
       String errorMessage = String.format("updateTransactionStatus:: status update from %s to %s is not implemented", currentStatus, requestedStatus);
       log.warn(errorMessage);
       throw new IllegalArgumentException(errorMessage);
     }
   }
 
+  public void cancelTransactionRequest(TransactionEntity transactionEntity){
+    try {
+      if(circulationService.cancelRequestIfExistOrNull(transactionEntity) == null){
+        cancelTransactionEntity(transactionEntity);
```
Comment (Vignesh-kalyanasundaram) [inline src/main/java/org/folio/dcb/service/impl/BaseLibraryService.java:122]: Is this condition is correct ?
 Looks like the method **cancelRequestIfExistOrNull** will return null only if the getCancellationRequestIfOpenOrNull method returns null. But we need to update the transaction status when the cancellation happens right ?

Comment (MagzhanArtykov) [reply src/main/java/org/folio/dcb/service/impl/BaseLibraryService.java]: State of transaction will be changed when kafka triggers cancel event from mod-circulation-storage

**Review by SinghAdes — APPROVED** [review]

**Review by Vignesh-kalyanasundaram — APPROVED** [review]

**Review by gurleenkaurbp — APPROVED** [review]

**Pull request merged**